File: Kalorie/Kalorie/Core/UseCases/DeleteMealTypeUseCase.swift
//
//  DeleteMealTypeUseCase.swift
//  Kalorie
//
//  Created by Josef Antoni on 29.06.2026.
//

import Foundation

protocol DeleteMealTypeUseCaseProtocol {
    func callAsFunction(_ mealType: MealTypeDomain) async throws
}

struct DeleteMealTypeUseCase: DeleteMealTypeUseCaseProtocol {

    // MARK: - Properties

    private let dataProvider: any FirestoreDataProviderProtocol
    private let authProvider: any AuthProviderProtocol

    // MARK: - Init

    init(dataProvider: any FirestoreDataProviderProtocol, authProvider: any AuthProviderProtocol) {
        self.dataProvider = dataProvider
        self.authProvider = authProvider
    }

    // MARK: - Functions

    func callAsFunction(_ mealType: MealTypeDomain) async throws {
        guard let userId = authProvider.userId else { throw AuthError.notAuthenticated }
        try await dataProvider.deleteAsync(id: mealType.id, from: Constants.Firestore.mealTypes(userId: userId))
    }
}

#if DEBUG
struct DeleteMealTypeUseCaseFake: DeleteMealTypeUseCaseProtocol {

    // MARK: - Properties

    var shouldThrow = false

    // MARK: - Functions

    func callAsFunction(_ mealType: MealTypeDomain) async throws {
        if shouldThrow { throw URLError(.unknown) }
    }
}
#endif

File: Android/app/src/main/java/antoni/kalorie/core/usecases/DeleteMealTypeUseCase.kt
package antoni.kalorie.core.usecases

import antoni.kalorie.core.auth.AuthError
import antoni.kalorie.core.auth.AuthProviderProtocol
import antoni.kalorie.core.models.MealTypeDomain
import antoni.kalorie.core.networking.FirestoreDataProviderProtocol
import antoni.kalorie.core.utils.Constants

interface DeleteMealTypeUseCaseProtocol {
    suspend operator fun invoke(mealType: MealTypeDomain)
}

class DeleteMealTypeUseCase(
    private val dataProvider: FirestoreDataProviderProtocol,
    private val authProvider: AuthProviderProtocol,
) : DeleteMealTypeUseCaseProtocol {

    // MARK: - Functions

    override suspend fun invoke(mealType: MealTypeDomain) {
        val userId = authProvider.userId ?: throw AuthError.NotAuthenticated
        dataProvider.deleteAsync(id = mealType.id, from = Constants.Firestore.mealTypes(userId))
    }
}
